mlflow/pyfunc/__init__.py

```diff
@@ -941,7 +1005,91 @@ def predict(*args):
         else:
             return result[result.columns[0]]
 
-    udf = pandas_udf(predict, result_type)
+    def predict(iterator):
+        # importing here to prevent circular import
+        from mlflow.pyfunc import scoring_server
+        from mlflow.pyfunc.scoring_server.client import ScoringServerClient, prepare_env
+
+        # Note: this is a pandas udf function in iteration style, which takes an iterator of
+        # tuple of pandas.Series and outputs an iterator of pandas.Series.
+
+        scoring_server_proc = None
+
+        # TODO: Support virtual env.
+        #
+        # TODO: For conda/virtualenv restored env cases,
+        #  For each individual python process (driver side), create individual and temporary
+        #  conda env dir / virtualenv env dir and when process exit,
+        #  delete the temporary env dir.
+        #  The reason is
+        #   1. env dir might be a large size directory and cleaning it when process exit
+        #      help saving disk space.
+        #   2. We have conda package cache dir and pip cache dir which are shared across all
+        #      python processes which help reducing downloading time.
+        #   3. Avoid race conditions related issues.
+        #
+        # TODO:
+        #   For NFS available case, set conda env dir / virtualenv env dir in sub-directory under
+        #   NFS directory, and in spark driver side prepare restored env once, and then all
+        #   spark UDF tasks running on spark workers can skip re-creating the restored env.
+        if env_manager == "conda":
+            server_port = find_free_port()
+
+            if should_use_spark_to_broadcast_file:
+                local_model_path_on_executor = _SparkDirectoryDistributor.get_or_extract(
+                    archive_path
+                )
+                # Call "prepare env" in advance in order to reduce scoring server launch time.
+                # So that we can use a shorter timeout when call `client.wait_server_ready`,
+                # otherwise we have to set a long timeout for `client.wait_server_ready` time,
+                # this prevents spark UDF task failing fast if other exception raised when scoring
+                # server launching.
+                prepare_env(local_model_path_on_executor)
+            else:
+                local_model_path_on_executor = local_model_path
+            # launch scoring server
+
+            # TODO: adjust timeout for server requests handler.
+            scoring_server_proc = _get_flavor_backend(
+                local_model_path_on_executor, no_conda=False, workers=1, install_mlflow=False
+            ).serve(
+                model_uri=local_model_path_on_executor,
+                port=server_port,
+                host='127.0.0.1',
+                enable_mlserver=False,
+                blocking=False
+            )
+
+            client = ScoringServerClient("127.0.0.1", server_port)
+            client.wait_server_ready(timeout=30)
```

Comment:
Is 30 seconds enough? I think we should wait for at least 90 seconds in case the model is large and takes a long time to load.